proxy/http2/Http2ConnectionState.cc

```diff
@@ -1454,45 +1453,36 @@ Http2ConnectionState::send_a_data_frame(Http2Stream *stream, size_t &payload_len
     return Http2SendDataFrameResult::ERROR;
   }
 
-  // Select appropriate payload length
-  if (_sm->is_read_avail_more_than(0)) {
-    // We only need to check for window size when there is a payload
-    if (window_size <= 0) {
-      Http2StreamDebug(this->ua_session, stream->get_id(), "No window");
-      return Http2SendDataFrameResult::NO_WINDOW;
-    }
-    // Copy into the payload buffer. Seems like we should be able to skip this copy step
-    payload_length = write_available_size;
-    payload_length = _sm->read(payload_buffer, static_cast<int64_t>(write_available_size));
-  } else {
-    payload_length = 0;
-  }
-
   // Are we at the end?
   // If we return here, we never send the END_STREAM in the case of a early terminating OS.
   // OK if there is no body yet. Otherwise continue on to send a DATA frame and delete the stream
-  if (!stream->is_body_done() && payload_length == 0) {
+  if (!stream->is_body_done() && !_sm->is_read_avail_more_than(0)) {
     Http2StreamDebug(this->ua_session, stream->get_id(), "No payload");
     return Http2SendDataFrameResult::NO_PAYLOAD;
+  } else if (_sm->is_read_avail_more_than(0)) {
+    // We only need to check for window size when there is a payload
+    if (window_size <= 0) {
+      Http2StreamDebug(this->ua_session, stream->get_id(), "No window");
+      return Http2SendDataFrameResult::NO_WINDOW;
+    }
   }
 
-  if (stream->is_body_done() && !_sm->is_read_avail_more_than(0)) {
+  if (stream->is_body_done() && !_sm->is_read_avail_more_than(write_available_size)) {
     flags |= HTTP2_FLAGS_DATA_END_STREAM;
   }
 
+  Http2Frame data(HTTP2_FRAME_TYPE_DATA, stream->get_id(), flags, _sm);
+
+  payload_length = std::min(static_cast<size_t>(_sm->read_avail()), write_available_size);
+  data.set_payload_size(payload_length);
+
   // Update window size
   this->decrement_client_rwnd(payload_length);
   stream->decrement_client_rwnd(payload_length);
 
-  // Create frame
   Http2StreamDebug(ua_session, stream->get_id(), "Send a DATA frame - client window con: %5zd stream: %5zd payload: %5zd",
                    _client_rwnd, stream->client_rwnd(), payload_length);
 
-  Http2Frame data(HTTP2_FRAME_TYPE_DATA, stream->get_id(), flags);
-  data.alloc(buffer_size_index[HTTP2_FRAME_TYPE_DATA]);
-  http2_write_data(payload_buffer, payload_length, data.write());
```

Comment:
This is the only place that uses `http2_write_data()`. Please remove the function definition too.